# Analyzing Neighborhoods in <font color=lightblue>Downtown Toronto</font> VS. <font color=Pink>Downtown Vancouver</font>

##  Project objective: Explore the similarities between Downtown Toronto and Downtown Vancouver by analyzing the popular venues in the neighbourhoods

#### Importing Libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
#!conda install -c conda-forge geopy --yes

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!pip install bs4
#!pip install plotly
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

#!pip install lxml
#!pip install html5lib

print('Libraries imported.')

Libraries imported.


# <font color=lightblue>Part 1: Analyzing Neighbourhoods in Downtown Toronto</font>

# Wiki Link URL

In [2]:
my_url1=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text   #Wiki page for Toronto Postal Codes

# 1. Create Soup Object from URL

In [3]:
soup1 = BeautifulSoup(my_url1,"html5lib")

In [4]:
tag_object1=soup1.title
print("tag object:",tag_object1)

tag object: <title>List of postal codes of Canada: M - Wikipedia</title>


In [5]:
table_contents=[]
table=soup1.find('table')
table

<table cellpadding="2" cellspacing="0" rules="all" style="width:100%; border-collapse:collapse; border:1px solid #ccc;">

<tbody><tr>
<td style="width:11%; vertical-align:top; color:#ccc;">
<p><b>M1A</b><br/><span style="font-size:85%;"><i>Not assigned</i></span>
</p>
</td>
<td style="width:11%; vertical-align:top; color:#ccc;">
<p><b>M2A</b><br/><span style="font-size:85%;"><i>Not assigned</i></span>
</p>
</td>
<td style="width:11%; vertical-align:top;">
<p><b>M3A</b><br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>)</span>
</p>
</td>
<td style="width:11%; vertical-align:top;">
<p><b>M4A</b><br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>)</span>
</p>
</td>
<td style="width:11%; vertical-align:top;">
<p><b>M5A</b><br/><span style="font-size:85%;"><a h

In [6]:
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)
        
print(table_contents)


[{'PostalCode': 'M3A', 'Borough': 'North York', 'Neighborhood': 'Parkwoods'}, {'PostalCode': 'M4A', 'Borough': 'North York', 'Neighborhood': 'Victoria Village'}, {'PostalCode': 'M5A', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Regent Park, Harbourfront'}, {'PostalCode': 'M6A', 'Borough': 'North York', 'Neighborhood': 'Lawrence Manor, Lawrence Heights'}, {'PostalCode': 'M7A', 'Borough': "Queen's Park", 'Neighborhood': 'Ontario Provincial Government'}, {'PostalCode': 'M9A', 'Borough': 'Etobicoke', 'Neighborhood': 'Islington Avenue'}, {'PostalCode': 'M1B', 'Borough': 'Scarborough', 'Neighborhood': 'Malvern, Rouge'}, {'PostalCode': 'M3B', 'Borough': 'North York', 'Neighborhood': 'Don Mills North'}, {'PostalCode': 'M4B', 'Borough': 'East York', 'Neighborhood': 'Parkview Hill, Woodbine Gardens'}, {'PostalCode': 'M5B', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Garden District, Ryerson'}, {'PostalCode': 'M6B', 'Borough': 'North York', 'Neighborhood': 'Glencairn'}, {'PostalCode': 'M9

# Generating the table with PostalCode, Borough, and Neighborhood.

In [7]:
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

## Shape of the table:

In [8]:
df.shape

(103, 3)

## 2. Import Coordinates data frame from csv file and join to original table on postal code to get Latitude and Longitude columns

In [9]:
data_cds = pd.read_csv('https://cocl.us/Geospatial_data') 
data_cds.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## Dataframe with Coordinates:

In [10]:
df_final = pd.merge(df, data_cds, left_on='PostalCode', right_on='Postal Code', how='left').drop('Postal Code', axis=1)
df_final.head(20)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   
15        M5C  Downtown Toronto   
16        M6C              York   
17        M9C         Etobicoke   
18        M1E       Scarborough   
19        M4E      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                       Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                     Don Mills North  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724  
12             Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
13                                    Don Mills South  43.725900 -79.340923  
14                                   Woodbine Heights  43.695344 -79.318389  
15                                     St. James Town  43.651494 -79.375418  
16                                 Humewood-Cedarvale  43.693781 -79.428191  
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201  
18                  Guildwood, Morningside, West Hill  43.763573 -79.188711  
19                                        The Beaches  43.676357 -79.293031

# 3. Clustering and Mapping

In [11]:
#pip install folium

In [12]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

In [13]:
trt_dt_data = df_final[df_final['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
trt_dt_data

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M5B  Downtown Toronto   
2         M5C  Downtown Toronto   
3         M5E  Downtown Toronto   
4         M5G  Downtown Toronto   
5         M6G  Downtown Toronto   
6         M5H  Downtown Toronto   
7         M5J  Downtown Toronto   
8         M5K  Downtown Toronto   
9         M5L  Downtown Toronto   
10        M5S  Downtown Toronto   
11        M5T  Downtown Toronto   
12        M5V  Downtown Toronto   
13        M4W  Downtown Toronto   
14        M4X  Downtown Toronto   
15        M5X  Downtown Toronto   
16        M4Y  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1                            Garden District, Ryerson  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         Berczy Park  43.644771 -79.373306  
4                                  Central Bay Street  43.657952 -79.387383  
5                                            Christie  43.669542 -79.422564  
6                            Richmond, Adelaide, King  43.650571 -79.384568  
7   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
8            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
9                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
10                     University of Toronto, Harbord  43.662696 -79.400049  
11          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
12  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
13                                           Rosedale  43.679563 -79.377529  
14                        St. James Town, Cabbagetown  43.667967 -79.367675  
15             First Canadian Place, Underground city  43.648429 -79.382280  
16                               Church and Wellesley  43.665860 -79.383160

## Visualize the Toronto Neighbourhood

In [14]:
#pip install Nominatim
from geopy.geocoders import Nominatim

In [15]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

# create map of Manhattan using latitude and longitude values
map_trt = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(trt_dt_data['Latitude'], trt_dt_data['Longitude'], trt_dt_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_trt)  
    
map_trt

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


## Define Foursquare Credentials and Version

In [16]:

CLIENT_ID = 'GL3GBNB0DRIQ3Z5QJE54ZHPPKZARCKPRBAKRH2NDF0C4DLP2' # your Foursquare ID
CLIENT_SECRET = 'FS1GSVHKHW4NKIWWUYC4XQISQ0GEOBHVHCP0XR5W2PCBG3FR' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)




Your credentails:
CLIENT_ID: GL3GBNB0DRIQ3Z5QJE54ZHPPKZARCKPRBAKRH2NDF0C4DLP2
CLIENT_SECRET:FS1GSVHKHW4NKIWWUYC4XQISQ0GEOBHVHCP0XR5W2PCBG3FR


In [17]:


trt_dt_data.loc[0, 'Neighborhood']


neighborhood_latitude = trt_dt_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = trt_dt_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = trt_dt_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))




Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


## Let's get the top 100 venues that are in Regent Park, Harbourfront within a radius of 500 meters

In [18]:
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url


results = requests.get(url).json()
results


{'meta': {'code': 200, 'requestId': '606e74f7a465a5165e55c7a1'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 44,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '53b8466a498e83df908c3f21',
       'name': 'Tandem Coffee',
       'location': {'address': '368 King St E',
        'crossStreet': 'at Trinity St',
        'lat': 43.65355870959944,
        'lng': -79.36180945913513,
        'labeledLatLngs': [{'label': 'display',
 

In [19]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']
    
    
venues = results['response']['groups'][0]['items']
    
nearby_venues = pd.json_normalize(venues) # flatten JSON


In [20]:

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))


44 venues were returned by Foursquare.


## Explore Neighborhoods in Toronto

In [21]:


def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)



trt_venues = getNearbyVenues(names=trt_dt_data['Neighborhood'],
                                   latitudes=trt_dt_data['Latitude'],
                                   longitudes=trt_dt_data['Longitude']
                                  )


print(trt_venues.shape)
trt_venues.head()


Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley
(1095, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0           Tandem Coffee       43.653559       -79.361809   
1        Roselle Desserts       43.653447       -79.362017   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0          Coffee Shop  
1               Bakery  
2  Distribution Center  
3           Restaurant  
4                  Spa

## Analyze Each Neighborhood

In [22]:


trt_venues.groupby('Neighborhood').count()

print('There are {} uniques categories.'.format(len(trt_venues['Venue Category'].unique())))


# one hot encoding
trt_onehot = pd.get_dummies(trt_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
trt_onehot['Neighborhood'] = trt_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [trt_onehot.columns[-1]] + list(trt_onehot.columns[:-1])
trt_onehot = trt_onehot[fixed_columns]

trt_onehot.head()




There are 205 uniques categories.


Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                   0          0           0           0       0     0    0   
1                   0          0           0           0       1     0    0   
2                   0          0           0           0       0     0    0   
3                   0          0           0           0       0     0    0   
4                   0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Candy Store  \
0             0              0        0     0            0   
1             0              0        0     0            0   
2             0              0        0     0            0   
3             0              0        0     0            0   
4             0              0        0     0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0            0 

In [23]:

trt_onehot.shape



(1095, 205)

In [24]:

trt_grouped = trt_onehot.groupby('Neighborhood').mean().reset_index()
trt_grouped

trt_grouped.shape



(17, 205)

## Print each neighborhood along with the top 5 most common venues

In [25]:

num_top_venues = 5

for hood in trt_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = trt_grouped[trt_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')
    
    


----Berczy Park----
            venue  freq
0     Coffee Shop  0.07
1          Bakery  0.05
2    Cocktail Bar  0.05
3  Farmers Market  0.03
4        Pharmacy  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2  Airport Terminal  0.12
3   Harbor / Marina  0.06
4               Bar  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1                Café  0.06
2      Sandwich Place  0.06
3  Italian Restaurant  0.05
4        Burger Joint  0.03


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  0.19
2           Park  0.12
3    Candy Store  0.06
4     Baby Store  0.06


----Church and Wellesley----
                  venue  freq
0           Coffee Shop  0.08
1      Sushi Restaurant  0.07
2   Japanese Restaurant  0.07
3            Restaurant  0.04
4  Fast Food Restaurant  0.0

In [26]:

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


## Create the new dataframe and display the top 10 venues for each neighborhood

In [27]:


num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = trt_grouped['Neighborhood']

for ind in np.arange(trt_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(trt_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar           Cheese Shop   
1       Airport Service      Airport Terminal           Coffee Shop   
2        Sandwich Place                  Café    Italian Restaurant   
3                  Café                  Park    Italian Restaurant   
4   Japanese Restaurant      Sushi Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
0              Pharmacy              Beer Bar            Farmers Market   
1                 Plane      Sculpture Garden             Boat or Ferry   
2          Burger Joint           Salad Place           Bubble Tea Shop   
3    Athletics & Sports            Restaurant               Candy Store   
4           Yoga Studio           Men's Store  Mediterranean Restaurant   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0            Restaurant     Seafood Restaurant           Liquor Store  
1   Rental Car Location                    Bar               Boutique  
2       Thai Restaurant  Portuguese Restaurant             Poke Place  
3            Baby Store              Nightclub            Coffee Shop  
4                 Hotel                Gay Bar   Fast Food Restaurant

## Run k-means to cluster the neighborhood into 5 clusters

In [28]:
# set number of clusters
kclusters = 5

trt_grouped_clustering = trt_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(trt_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 


# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

trt_merged = trt_dt_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
trt_merged = trt_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

trt_merged.head() # check the last columns!



PostalCode           Borough               Neighborhood   Latitude  \
0        M5A  Downtown Toronto  Regent Park, Harbourfront  43.654260   
1        M5B  Downtown Toronto   Garden District, Ryerson  43.657162   
2        M5C  Downtown Toronto             St. James Town  43.651494   
3        M5E  Downtown Toronto                Berczy Park  43.644771   
4        M5G  Downtown Toronto         Central Bay Street  43.657952   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               0           Coffee Shop                   Pub   
1 -79.378937               0        Clothing Store           Coffee Shop   
2 -79.375418               0           Coffee Shop                  Café   
3 -79.373306               0           Coffee Shop                Bakery   
4 -79.387383               0           Coffee Shop        Sandwich Place   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                Bakery                  Park             Breakfast Spot   
1       Bubble Tea Shop                  Café  Middle Eastern Restaurant   
2        Cosmetics Shop          Cocktail Bar             Farmers Market   
3          Cocktail Bar           Cheese Shop                   Pharmacy   
4                  Café    Italian Restaurant               Burger Joint   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                  Café               Theater  Gym / Fitness Center   
1        Cosmetics Shop                 Hotel      Ramen Restaurant   
2                  Park            Restaurant   American Restaurant   
3              Beer Bar        Farmers Market            Restaurant   
4           Salad Place       Bubble Tea Shop       Thai Restaurant   

   9th Most Common Venue 10th Most Common Venue  
0            Event Space  Performing Arts Venue  
1              Bookstore           Burger Joint  
2     Italian Restaurant               Creperie  
3     Seafood Restaurant           Liquor Store  
4  Portuguese Restaurant             Poke Place

## Visualize the resulting clusters

In [29]:

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(trt_merged['Latitude'], trt_merged['Longitude'], trt_merged['Neighborhood'], trt_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine the Five Clusters

In [30]:
trt_merged.loc[trt_merged['Cluster Labels'] == 0, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0        Clothing Store   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0                  Café   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                    Pub                Bakery                   Park   
1            Coffee Shop       Bubble Tea Shop                   Café   
2                   Café        Cosmetics Shop           Cocktail Bar   
3                 Bakery          Cocktail Bar            Cheese Shop   
4         Sandwich Place                  Café     Italian Restaurant   
6                   Café            Restaurant          Deli / Bodega   
7               Aquarium                 Hotel                   Café   
8                  Hotel                  Café     Seafood Restaurant   
9             Restaurant                  Café                  Hotel   
11           Coffee Shop    Mexican Restaurant  Vietnamese Restaurant   
14                Bakery                  Café     Italian Restaurant   
15                  Café                 Hotel                    Gym   
16   Japanese Restaurant      Sushi Restaurant             Restaurant   

            5th Most Common Venue 6th Most Common Venue  \
0                  Breakfast Spot                  Café   
1       Middle Eastern Restaurant        Cosmetics Shop   
2                  Farmers Market                  Park   
3                        Pharmacy              Beer Bar   
4                    Burger Joint           Salad Place   
6                  Clothing Store                   Gym   
7              Italian Restaurant   Sporting Goods Shop   
8              Italian Restaurant            Restaurant   
9              Italian Restaurant                   Gym   
11  Vegetarian / Vegan Restaurant                  Park   
14                           Park            Restaurant   
15            Japanese Restaurant            Restaurant   
16                    Yoga Studio           Men's Store   

       7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
0                    Theater  Gym / Fitness Center            Event Space   
1                      Hotel      Ramen Restaurant              Bookstore   
2                 Restaurant   American Restaurant     Italian Restaurant   
3             Farmers Market            Restaurant     Seafood Restaurant   
4            Bubble Tea Shop       Thai Restaurant  Portuguese Restaurant   
6            Thai Restaurant                 Hotel         Cosmetics Shop   
7             Scenic Lookout               Brewery    Fried Chicken Joint   
8        Japanese Restaurant           Salad Place    Sporting Goods Shop   
9         Seafood Restaurant         Deli / Bodega    Japanese Restaurant   
11       Arts & Crafts Store           Gaming Cafe                 Bakery   
14                       Pub           Pizza Place                  Plaza   
15          Asian Restaurant         Deli / Bodega     Seafood Restaurant   
16  Mediterranean Restaurant                 Hotel                Gay Bar   

   10th Most Common Venue  
0   Performing Arts Venue  
1            Burger Joint  
2                Creperie  
3            Liquor Store  
4              Poke Place  
6        Sushi Restaurant  
7              Rest

In [31]:
trt_merged.loc[trt_merged['Cluster Labels'] == 1, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13            Playground                 Trail      Department Store   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
13     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13            Donut Shop      Doner Restaurant                Dog Run

In [32]:
trt_merged.loc[trt_merged['Cluster Labels'] == 2, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5  Downtown Toronto               2         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5                  Café                  Park    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5    Athletics & Sports            Restaurant           Candy Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5            Baby Store             Nightclub            Coffee Shop

In [33]:
trt_merged.loc[trt_merged['Cluster Labels'] == 3, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12  Downtown Toronto               3        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12       Airport Service      Airport Terminal           Coffee Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12                 Plane      Sculpture Garden         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12   Rental Car Location                   Bar               Boutique

In [34]:
trt_merged.loc[trt_merged['Cluster Labels'] == 4, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10  Downtown Toronto               4                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10             Bookstore                   Bar    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10   Japanese Restaurant        Sandwich Place                Bakery   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
10           Yoga Studio              Beer Bar             Beer Store

# <font color=pink>Part 2: Analyzing Neighbourhoods in Downtown Vancouver</font>

In [35]:
my_url2=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_V').text   #Wiki page for Vancouver Postal Codes

In [36]:
soup2 = BeautifulSoup(my_url2,"html5lib")

In [37]:
tag_object2=soup2.title
print("tag object:",tag_object2)

tag object: <title>List of postal codes of Canada: V - Wikipedia</title>


In [38]:
table_contents2=[]
table2=soup2.find('table')
table2

<table class="wikitable sortable">
<tbody><tr>
<td valign="top" width="11.1%"><b>V1A</b><br/><span style="font-size: smaller; line-height: 125%;"><a href="/wiki/Kimberley,_British_Columbia" title="Kimberley, British Columbia">Kimberley</a></span>
</td>
<td valign="top" width="11.1%"><b>V2A</b><br/><span style="font-size: smaller; line-height: 125%;"><a href="/wiki/Penticton" title="Penticton">Penticton</a></span>
</td>
<td valign="top" width="11.1%"><b>V3A</b><br/><span style="font-size: smaller; line-height: 125%;"><a href="/wiki/Langley,_British_Columbia_(district_municipality)" title="Langley, British Columbia (district municipality)">Langley Township</a><br/>(Langley City)</span>
</td>
<td valign="top" width="11.1%"><b>V4A</b><br/><span style="font-size: smaller; line-height: 125%;"><a href="/wiki/Surrey,_British_Columbia" title="Surrey, British Columbia">Surrey</a><br/>Southwest</span>
</td>
<td valign="top" width="11.1%"><b>V5A</b><br/><span style="font-size: smaller; line-height

In [39]:
for row in table2.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.text[:3]
        cell['City'] = (row.span.text).split('(')[0]
        table_contents2.append(cell)
        
print(table_contents2)


[{'PostalCode': 'V1A', 'City': 'Kimberley'}, {'PostalCode': 'V2A', 'City': 'Penticton'}, {'PostalCode': 'V3A', 'City': 'Langley Township'}, {'PostalCode': 'V4A', 'City': 'SurreySouthwest'}, {'PostalCode': 'V5A', 'City': 'Burnaby'}, {'PostalCode': 'V6A', 'City': 'Vancouver'}, {'PostalCode': 'V7A', 'City': 'RichmondSouth'}, {'PostalCode': 'V8A', 'City': 'Powell River'}, {'PostalCode': 'V9A', 'City': 'Victoria'}, {'PostalCode': 'V1B', 'City': 'VernonEast'}, {'PostalCode': 'V2B', 'City': 'KamloopsNorthwest'}, {'PostalCode': 'V3B', 'City': 'Port CoquitlamCentral'}, {'PostalCode': 'V4B', 'City': 'White Rock'}, {'PostalCode': 'V5B', 'City': 'Burnaby'}, {'PostalCode': 'V6B', 'City': 'Vancouver'}, {'PostalCode': 'V7B', 'City': 'Richmond'}, {'PostalCode': 'V8B', 'City': 'Squamish'}, {'PostalCode': 'V9B', 'City': 'Victoria'}, {'PostalCode': 'V1C', 'City': 'Cranbrook'}, {'PostalCode': 'V2C', 'City': 'KamloopsCentral and Southeast'}, {'PostalCode': 'V3C', 'City': 'Port CoquitlamSouth'}, {'PostalCod

In [40]:
df2=pd.DataFrame(table_contents2)
df2

PostalCode                           City
0          V1A                      Kimberley
1          V2A                      Penticton
2          V3A               Langley Township
3          V4A                SurreySouthwest
4          V5A                        Burnaby
5          V6A                      Vancouver
6          V7A                  RichmondSouth
7          V8A                   Powell River
8          V9A                       Victoria
9          V1B                     VernonEast
10         V2B              KamloopsNorthwest
11         V3B          Port CoquitlamCentral
12         V4B                     White Rock
13         V5B                        Burnaby
14         V6B                      Vancouver
15         V7B                       Richmond
16         V8B                       Squamish
17         V9B                       Victoria
18         V1C                      Cranbrook
19         V2C  KamloopsCentral and Southeast
20         V3C            Port CoquitlamSouth
21         V4C                 DeltaNortheast
22         V5C                        Burnaby
23         V6C                      Vancouver
24         V7C              RichmondNorthwest
25         V8C                        Kitimat
26         V9C                       Victoria
27         V1E                     Salmon Arm
28         V2E         KamloopsSouth and West
29         V3E                 CoquitlamNorth
30         V4E                      DeltaEast
31         V5E                        Burnaby
32         V6E                      Vancouver
33         V7E              RichmondSouthwest
34         V8E                       Whistler
35         V9E                       Victoria
36         V1G                   Dawson Creek
37         V2G                  Williams Lake
38         V3G                 AbbotsfordEast
39         V4G              DeltaEast Central
40         V5G                        Burnaby
41         V6G                      Vancouver
42         V7G               North Vancouver 
43         V8G                        Terrace
44         V9G                      Ladysmith
45         V1H                     VernonWest
46         V2H                  KamloopsNorth
47         V3H                     Port Moody
48         V5H                        Burnaby
49         V6H                      Vancouver
50         V7H               North Vancouver 
51         V9H        Campbell RiverOutskirts
52         V1J                  Fort St. John
53         V2J                        Quesnel
54         V3J                 CoquitlamNorth
55         V5J                        Burnaby
56         V6J                      Vancouver
57         V7J               North Vancouver 
58         V8J                  Prince Rupert
59         V9J    CourtenayNorthern Outskirts
60         V1K                        Merritt
61         V2K             Prince GeorgeNorth
62         V3K                 CoquitlamSouth
63         V4K                 DeltaNorthwest
64         V5K                      Vancouver
65         V6K                      Vancouver
66         V7K               North Vancouver 
67         V8K             Salt Spring Island
68         V9K                 Qualicum Beach
69         V1L                         Nelson
70         V2L      Prince GeorgeEast Central
71         V3L       New WestminsterNortheast
72         V4L                 DeltaSoutheast
73         V5L                      Vancouver
74         V6L                      Vancouver
75         V7L               North Vancouver 
76         V8L                         Sidney
77         V9L                         Duncan
78         V1M          Langley TownshipNorth
79         V2M      Prince GeorgeWest Central
80         V3M       New WestminsterSouthwest
81         V4M                 DeltaSouthwest
82         V5M                      Vancouver
83         V6M                      Vancouver
84         V7M               North Vancouver 
85         V8M                Central Saanich
86

In [41]:
df_van = df2.loc[df2['City'] == 'Vancouver'] #only choose the postal codes for Vancouver
df_van = df_van.head(20)
df_van

PostalCode       City
5          V6A  Vancouver
14         V6B  Vancouver
23         V6C  Vancouver
32         V6E  Vancouver
41         V6G  Vancouver
49         V6H  Vancouver
56         V6J  Vancouver
64         V5K  Vancouver
65         V6K  Vancouver
73         V5L  Vancouver
74         V6L  Vancouver
82         V5M  Vancouver
83         V6M  Vancouver
91         V5N  Vancouver
92         V6N  Vancouver
99         V5P  Vancouver
100        V6P  Vancouver
108        V5R  Vancouver
109        V6R  Vancouver
117        V5S  Vancouver

## Shape of the table:

In [42]:
df_van.shape

(20, 2)

In [43]:
#pip install geocoder

## 2. Get coordinates based on postal codes using Geopy

In [44]:
import geocoder # import geocoder
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values


ModuleNotFoundError: No module named 'geocoder'

## Dataframe with Coordinates:

In [46]:
lat = []
lon=[]
for i in df_van['PostalCode']:
    address = '{}'.format(i)
    geolocator = Nominatim(user_agent="ny_explorer")
    location = geolocator.geocode(address)
    latitude = location.latitude
    longitude = location.longitude
    lat.append('{}'.format(latitude))
    lon.append('{}'.format(longitude))
df_van['Latitude'] = lat
df_van['Longitude'] = lon
df_van

PostalCode       City            Latitude            Longitude
5          V6A  Vancouver   49.27377017128552  -123.09947872733053
14         V6B  Vancouver   49.27811172378708  -123.11959791158985
23         V6C  Vancouver  49.287716588954346  -123.11519411838533
32         V6E  Vancouver   49.28801282544443  -123.12108056609158
41         V6G  Vancouver          49.3002702  -123.13779663860902
49         V6H  Vancouver   49.25680006013691   -123.1331282550357
56         V6J  Vancouver   49.26091372148124  -123.14577875870242
64         V5K  Vancouver   49.28171754656246   -123.0400063294856
65         V6K  Vancouver   49.26895274770836  -123.16501916876759
73         V5L  Vancouver  49.280200918758624  -123.06656328873324
74         V6L  Vancouver    49.2534450006267  -123.17202047499656
82         V5M  Vancouver   49.26007804182507   -123.0401599764032
83         V6M  Vancouver   49.23445007981276  -123.14541062858737
91         V5N  Vancouver    49.2646254249102  -123.06586883849849
92         V6N  Vancouver  49.224594327808234   -123.1686106806456
99         V5P  Vancouver   49.22275069744015  -123.06755955330607
100        V6P  Vancouver  49.218349248173894  -123.14992456434537
108        V5R  Vancouver  49.239822993458404  -123.04080030032974
109        V6R  Vancouver   49.26911809217475  -123.19719593692786
117        V5S  Vancouver   49.20689898727592  -123.04459535837061

# 3. Clustering and Mapping

In [47]:
Vancouver_dt_data = df_van.reset_index(drop=True)
Vancouver_dt_data

PostalCode       City            Latitude            Longitude
0         V6A  Vancouver   49.27377017128552  -123.09947872733053
1         V6B  Vancouver   49.27811172378708  -123.11959791158985
2         V6C  Vancouver  49.287716588954346  -123.11519411838533
3         V6E  Vancouver   49.28801282544443  -123.12108056609158
4         V6G  Vancouver          49.3002702  -123.13779663860902
5         V6H  Vancouver   49.25680006013691   -123.1331282550357
6         V6J  Vancouver   49.26091372148124  -123.14577875870242
7         V5K  Vancouver   49.28171754656246   -123.0400063294856
8         V6K  Vancouver   49.26895274770836  -123.16501916876759
9         V5L  Vancouver  49.280200918758624  -123.06656328873324
10        V6L  Vancouver    49.2534450006267  -123.17202047499656
11        V5M  Vancouver   49.26007804182507   -123.0401599764032
12        V6M  Vancouver   49.23445007981276  -123.14541062858737
13        V5N  Vancouver    49.2646254249102  -123.06586883849849
14        V6N  Vancouver  49.224594327808234   -123.1686106806456
15        V5P  Vancouver   49.22275069744015  -123.06755955330607
16        V6P  Vancouver  49.218349248173894  -123.14992456434537
17        V5R  Vancouver  49.239822993458404  -123.04080030032974
18        V6R  Vancouver   49.26911809217475  -123.19719593692786
19        V5S  Vancouver   49.20689898727592  -123.04459535837061

## Visualize the Vancouver Neighbourhood

In [55]:
address = 'Vancouver, BC'

geolocator = Nominatim(user_agent="Vancouver_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Vancouver are {}, {}.'.format(latitude, longitude))

# create map of Manhattan using latitude and longitude values
map_vcv = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Vancouver_dt_data['Latitude'], Vancouver_dt_data['Longitude'], Vancouver_dt_data['PostalCode']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_vcv)  
    
map_vcv

The geograpical coordinate of Vancouver are 49.2608724, -123.1139529.


## Explore nearby venues of downtown Vancouver

In [56]:
vcv_venues = getNearbyVenues(names=Vancouver_dt_data['PostalCode'],
                                   latitudes=Vancouver_dt_data['Latitude'],
                                   longitudes=Vancouver_dt_data['Longitude']
                                  )


print(vcv_venues.shape)
vcv_venues.head()

V6A
V6B
V6C
V6E
V6G
V6H
V6J
V5K
V6K
V5L
V6L
V5M
V6M
V5N
V6N
V5P
V6P
V5R
V6R
V5S
(529, 7)


Neighborhood Neighborhood Latitude Neighborhood Longitude  \
0          V6A     49.27377017128552    -123.09947872733053   
1          V6A     49.27377017128552    -123.09947872733053   
2          V6A     49.27377017128552    -123.09947872733053   
3          V6A     49.27377017128552    -123.09947872733053   
4          V6A     49.27377017128552    -123.09947872733053   

                                     Venue  Venue Latitude  Venue Longitude  \
0                           Creekside Park       49.274641      -123.102701   
1                                   Boxcar       49.276613      -123.100076   
2                          Pizzeria Farina       49.276636      -123.099967   
3                                 Torafuku       49.275951      -123.099814   
4  Science World at TELUS World of Science       49.273310      -123.103501   

     Venue Category  
0              Park  
1               Bar  
2       Pizza Place  
3  Asian Restaurant  
4    Science Museum

## Analyze Each Neighborhood

In [59]:
vcv_venues.groupby('Neighborhood').count()

print('There are {} uniques categories.'.format(len(vcv_venues['Venue Category'].unique())))


# one hot encoding
vcv_onehot = pd.get_dummies(vcv_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
vcv_onehot['Neighborhood'] = vcv_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [vcv_onehot.columns[-1]] + list(vcv_onehot.columns[:-1])
vcv_onehot = vcv_onehot[fixed_columns]

vcv_onehot.head(15)

There are 169 uniques categories.


Neighborhood  Accessories Store  Airport Terminal  American Restaurant  \
0           V6A                  0                 0                    0   
1           V6A                  0                 0                    0   
2           V6A                  0                 0                    0   
3           V6A                  0                 0                    0   
4           V6A                  0                 0                    0   
5           V6A                  0                 0                    0   
6           V6A                  0                 0                    0   
7           V6A                  0                 0                    0   
8           V6A                  0                 0                    0   
9           V6A                  0                 0                    0   
10          V6A                  0                 0                    0   
11          V6A                  0                 0                    0   
12          V6A                  0                 0                    0   
13          V6A                  0                 0                    0   
14          V6A                  0                 0                    0   

    Amphitheater  Aquarium  Art Gallery  Arts & Crafts Store  \
0              0         0            0                    0   
1              0         0            0                    0   
2              0         0            0                    0   
3              0         0            0                    0   
4              0         0            0                    0   
5              0         0            0                    0   
6              0         0            0                    0   
7              0         0            0                    0   
8              0         0            0                    0   
9              0         0            0                    0   
10             0         0            0                    0   
11             0         0            0                    0   
12             0         0            0                    0   
13             0         0            0                    0   
14             0         0            0                    0   

    Asian Restaurant  Athletics & Sports  Australian Restaurant  Bakery  Bank  \
0                  0                   0                      0       0     0   
1                  0                   0                      0       0     0   
2                  0                   0                      0       0     0   
3                  1                   0                      0       0     0   
4                  0                   0                      0       0     0   
5                  0                   0                      0       0     0   
6                  0                   0                      0       0     0   
7                  0                   0                      0       0     0   
8                  0                   0                      0       0     0   
9                  0                   0                      0       0     0   
10                 0                   0                      0       0     0   
11                 0                   0                      0       0     0   
12                 0                   0                      0       0     0   
13                 0                   0                      0       0     0   
14                 0                   0                      0       0     0   

    Bar  Baseball Field  Beach  Beer Garden  Boat or Ferry  Bookstore  \
0     0               0      0            0              0          0   
1     1               0      0            0              0          0   
2     0               0      0            0              0          0   
3     0               0      0            0              0          0   
4     0               0      0            0              0          0   
5     0   

In [60]:

vcv_onehot.shape



(529, 170)

In [61]:

vcv_grouped = vcv_onehot.groupby('Neighborhood').mean().reset_index()
vcv_grouped

vcv_grouped.shape



(20, 170)

## Print each neighborhood along with the top 5 most common venues

In [62]:

num_top_venues = 5

for hood in vcv_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = vcv_grouped[vcv_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')
    
    


----V5K----
                          venue  freq
0                   Event Space  0.08
1                    Theme Park  0.08
2  Theme Park Ride / Attraction  0.08
3              Sushi Restaurant  0.04
4                   Beer Garden  0.04


----V5L----
                venue  freq
0  Chinese Restaurant  0.10
1              Bakery  0.10
2             Brewery  0.10
3         Coffee Shop  0.07
4                Café  0.07


----V5M----
          venue  freq
0      Bus Stop  0.09
1   Pizza Place  0.09
2   Coffee Shop  0.09
3    Restaurant  0.05
4  Liquor Store  0.05


----V5N----
                  venue  freq
0           Coffee Shop  0.14
1      Sushi Restaurant  0.12
2           Pizza Place  0.07
3  Ethiopian Restaurant  0.05
4         Grocery Store  0.05


----V5P----
             venue  freq
0      Pizza Place  0.14
1      Post Office  0.05
2     Liquor Store  0.05
3     Noodle House  0.05
4  Motorcycle Shop  0.05


----V5R----
               venue  freq
0                Bar  0.17
1     

## Create the new dataframe and display the top 10 venues for each neighborhood

In [67]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted2 = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted2['Neighborhood'] = vcv_grouped['Neighborhood']

for ind in np.arange(trt_grouped.shape[0]):
    neighborhoods_venues_sorted2.iloc[ind, 1:] = return_most_common_venues(vcv_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted2.head()

Neighborhood 1st Most Common Venue         2nd Most Common Venue  \
0          V5K           Event Space  Theme Park Ride / Attraction   
1          V5L    Chinese Restaurant                       Brewery   
2          V5M           Pizza Place                   Coffee Shop   
3          V5N           Coffee Shop              Sushi Restaurant   
4          V5P           Pizza Place             Convenience Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Theme Park                Bakery              Bus Line   
1                Bakery           Coffee Shop                  Café   
2              Bus Stop                Office    Mexican Restaurant   
3           Pizza Place        Sandwich Place    Chinese Restaurant   
4                  Bank                  Park           Post Office   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                Market          Soccer Field                  Park   
1               Theater     Convenience Store       Thai Restaurant   
2         Metro Station             Pet Store    Chinese Restaurant   
3          Burger Joint         Grocery Store  Ethiopian Restaurant   
4          Noodle House            Restaurant        Sandwich Place   

  9th Most Common Venue     10th Most Common Venue  
0                  Farm                    Stadium  
1     Electronics Store              Deli / Bodega  
2          Liquor Store             Sandwich Place  
3           Music Venue                Record Shop  
4       Motorcycle Shop  Middle Eastern Restaurant

## Run k-means to cluster the neighborhood into 5 clusters

In [90]:
# set number of clusters
kclusters = 5

vcv_grouped_clustering = vcv_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(vcv_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

# add clustering labels
neighborhoods_venues_sorted2.insert(0, 'Cluster Labels', kmeans.labels_)

#vcv_merged = Vancouver_dt_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
#vcv_merged = vcv_merged.join(neighborhoods_venues_sorted2.set_index('Neighborhood'), on='Neighborhood')

#vcv_merged.head() # check the last columns!

neighborhoods_venues_sorted2.head()



Cluster Labels11  Cluster Labels1  Cluster Labels Neighborhood  \
0                 0                0               0          V5K   
1                 0                0               0          V5L   
2                 0                0               0          V5M   
3                 0                0               0          V5N   
4                 0                0               0          V5P   

  1st Most Common Venue         2nd Most Common Venue 3rd Most Common Venue  \
0           Event Space  Theme Park Ride / Attraction            Theme Park   
1    Chinese Restaurant                       Brewery                Bakery   
2           Pizza Place                   Coffee Shop              Bus Stop   
3           Coffee Shop              Sushi Restaurant           Pizza Place   
4           Pizza Place             Convenience Store                  Bank   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                Bakery              Bus Line                Market   
1           Coffee Shop                  Café               Theater   
2                Office    Mexican Restaurant         Metro Station   
3        Sandwich Place    Chinese Restaurant          Burger Joint   
4                  Park           Post Office          Noodle House   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0          Soccer Field                  Park                  Farm   
1     Convenience Store       Thai Restaurant     Electronics Store   
2             Pet Store    Chinese Restaurant          Liquor Store   
3         Grocery Store  Ethiopian Restaurant           Music Venue   
4            Restaurant        Sandwich Place       Motorcycle Shop   

      10th Most Common Venue  
0                    Stadium  
1              Deli / Bodega  
2             Sandwich Place  
3                Record Shop  
4  Middle Eastern Restaurant

In [100]:
#Join two tables
vcv_merged = Vancouver_dt_data.join(neighborhoods_venues_sorted2)
vcv_merged

PostalCode       City            Latitude            Longitude  \
0         V6A  Vancouver   49.27377017128552  -123.09947872733053   
1         V6B  Vancouver   49.27811172378708  -123.11959791158985   
2         V6C  Vancouver  49.287716588954346  -123.11519411838533   
3         V6E  Vancouver   49.28801282544443  -123.12108056609158   
4         V6G  Vancouver          49.3002702  -123.13779663860902   
5         V6H  Vancouver   49.25680006013691   -123.1331282550357   
6         V6J  Vancouver   49.26091372148124  -123.14577875870242   
7         V5K  Vancouver   49.28171754656246   -123.0400063294856   
8         V6K  Vancouver   49.26895274770836  -123.16501916876759   
9         V5L  Vancouver  49.280200918758624  -123.06656328873324   
10        V6L  Vancouver    49.2534450006267  -123.17202047499656   
11        V5M  Vancouver   49.26007804182507   -123.0401599764032   
12        V6M  Vancouver   49.23445007981276  -123.14541062858737   
13        V5N  Vancouver    49.2646254249102  -123.06586883849849   
14        V6N  Vancouver  49.224594327808234   -123.1686106806456   
15        V5P  Vancouver   49.22275069744015  -123.06755955330607   
16        V6P  Vancouver  49.218349248173894  -123.14992456434537   
17        V5R  Vancouver  49.239822993458404  -123.04080030032974   
18        V6R  Vancouver   49.26911809217475  -123.19719593692786   
19        V5S  Vancouver   49.20689898727592  -123.04459535837061   

    Cluster Labels11  Cluster Labels1  Cluster Labels Neighborhood  \
0                  0                0               0          V5K   
1                  0                0               0          V5L   
2                  0                0               0          V5M   
3                  0                0               0          V5N   
4                  0                0               0          V5P   
5                  4                4               4          V5R   
6                  4                4               4          V5S   
7                  0                0               0          V6A   
8                  0                0               0          V6B   
9                  0                0               0          V6C   
10                 0                0               0          V6E   
11                 0                0               0          V6G   
12                 0                0               0          V6H   
13                 0                0               0          V6J   
14                 0                0               0          V6K   
15                 1                1               1          V6L   
16                 0                0               0          V6M   
17                 2                2               2          V6N   
18                 0                0               0          V6P   
19                 3                3               3          V6R   

   1st Most Common Venue         2nd Most Common Venue  \
0            Event Space  Theme Park Ride / Attraction   
1     Chinese Restaurant                       Brewery   
2            Pizza Place                   Coffee Shop   
3            Coffee Shop              Sushi Restaurant   
4            Pizza Place             Convenience Store   
5                    Bar                         Hotel   
6             Print Shop                          Park   
7           Dessert Shop                          Park   
8                  Hotel           Japanese Restaurant   
9          Boat or Ferry                         Hotel   
10                 Hotel           American Restaurant   
11              Aquarium                         Trail   
12        Breakfast Spot                        Bakery   
13           Coffee Shop             Electronics Store   
14           Coffee Shop                          Park   
15    Italian Restaurant          Caribbean Restaurant   
16              Bus Stop            Chinese Restaurant   
17                   NaN              

## Visualize the resulting clusters

In [92]:

# create map
map_clusters2 = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(vcv_merged['Latitude'], vcv_merged['Longitude'], vcv_merged['Neighborhood'], vcv_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters2)
       
map_clusters2

## Examine the Five Clusters

In [93]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 0, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City  Cluster Labels1  Cluster Labels Neighborhood  \
0   Vancouver                0               0          V5K   
1   Vancouver                0               0          V5L   
2   Vancouver                0               0          V5M   
3   Vancouver                0               0          V5N   
4   Vancouver                0               0          V5P   
7   Vancouver                0               0          V6A   
8   Vancouver                0               0          V6B   
9   Vancouver                0               0          V6C   
10  Vancouver                0               0          V6E   
11  Vancouver                0               0          V6G   
12  Vancouver                0               0          V6H   
13  Vancouver                0               0          V6J   
14  Vancouver                0               0          V6K   
16  Vancouver                0               0          V6M   
18  Vancouver                0               0          V6P   

   1st Most Common Venue         2nd Most Common Venue  \
0            Event Space  Theme Park Ride / Attraction   
1     Chinese Restaurant                       Brewery   
2            Pizza Place                   Coffee Shop   
3            Coffee Shop              Sushi Restaurant   
4            Pizza Place             Convenience Store   
7           Dessert Shop                          Park   
8                  Hotel           Japanese Restaurant   
9          Boat or Ferry                         Hotel   
10                 Hotel           American Restaurant   
11              Aquarium                         Trail   
12        Breakfast Spot                        Bakery   
13           Coffee Shop             Electronics Store   
14           Coffee Shop                          Park   
16              Bus Stop            Chinese Restaurant   
18                   NaN                           NaN   

           3rd Most Common Venue            4th Most Common Venue  \
0                     Theme Park                           Bakery   
1                         Bakery                      Coffee Shop   
2                       Bus Stop                           Office   
3                    Pizza Place                   Sandwich Place   
4                           Bank                             Park   
7                            Bar                Korean Restaurant   
8                     Restaurant               Seafood Restaurant   
9                     Restaurant                      Coffee Shop   
10                    Restaurant                             Park   
11  Theme Park Ride / Attraction                      Zoo Exhibit   
12                          Park                      Coffee Shop   
13                   Yoga Studio              Japanese Restaurant   
14                          Café  Southern / Soul Food Restaurant   
16              Sushi Restaurant                 Asian Restaurant   
18                           NaN                              NaN   

   5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
0               Bus Line                  Market          Soccer Field   
1                   Café                 Theater     Convenience Store   
2     Mexican Restaurant           Metro Station             Pet Store   
3     Chinese Restaurant            Burger Joint         Grocery Store   
4            Post Office            Noodle House            Restaurant   
7            Pizza Place        Community Center           Coffee Shop   
8     Italian Restaurant             Coffee Shop                   Spa   
9                   Café               Hotel Bar                 Plaza   
10            Steakhouse      Miscellaneous Shop          Dessert Shop   
11     Outdoor Sculpture              Playground                  Café   
12             Pet Store              Shoe Store          Burger Joint   
13                   Bar  Furniture / Home Store                  Bank   
14  

In [95]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 1, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City  Cluster Labels1  Cluster Labels Neighborhood  \
15  Vancouver                1               1          V6L   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
15    Italian Restaurant  Caribbean Restaurant        Baseball Field   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
15           Zoo Exhibit                  Farm     Food & Drink Shop   

   7th Most Common Venue 8th Most Common Venue       9th Most Common Venue  \
15           Fish Market     Fish & Chips Shop  Financial or Legal Service   

   10th Most Common Venue  
15   Fast Food Restaurant

In [97]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 2, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City  Cluster Labels1  Cluster Labels Neighborhood  \
17  Vancouver                2               2          V6N   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
17                   NaN                   NaN                   NaN   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
17                   NaN                   NaN                   NaN   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
17                   NaN                   NaN                   NaN   

   10th Most Common Venue  
17                    NaN

In [98]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 3, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City  Cluster Labels1  Cluster Labels Neighborhood  \
19  Vancouver                3               3          V6R   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
19                   NaN                   NaN                   NaN   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
19                   NaN                   NaN                   NaN   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
19                   NaN                   NaN                   NaN   

   10th Most Common Venue  
19                    NaN

In [99]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 4, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City  Cluster Labels1  Cluster Labels Neighborhood  \
5  Vancouver                4               4          V5R   
6  Vancouver                4               4          V5S   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
5                   Bar                 Hotel              Bus Stop   
6            Print Shop                  Park              Bus Stop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
5     Fish & Chips Shop      Asian Restaurant                  Park   
6        Farmers Market          Burger Joint            Donut Shop   

  7th Most Common Venue 8th Most Common Venue       9th Most Common Venue  \
5           Zoo Exhibit           Fish Market  Financial or Legal Service   
6     Electronics Store  Ethiopian Restaurant                    Dive Bar   

  10th Most Common Venue  
5   Fast Food Restaurant  
6            Event Space

In [115]:
print(vcv_merged['1st Most Common Venue'].value_counts(normalize=True))
print(trt_merged['1st Most Common Venue'].value_counts(normalize=True))

Coffee Shop           0.176471
Hotel                 0.117647
Pizza Place           0.117647
Italian Restaurant    0.058824
Chinese Restaurant    0.058824
Aquarium              0.058824
Print Shop            0.058824
Dessert Shop          0.058824
Event Space           0.058824
Boat or Ferry         0.058824
Bar                   0.058824
Bus Stop              0.058824
Breakfast Spot        0.058824
Name: 1st Most Common Venue, dtype: float64
Coffee Shop       0.647059
Café              0.117647
Airport Lounge    0.058824
Grocery Store     0.058824
Park              0.058824
Clothing Store    0.058824
Name: 1st Most Common Venue, dtype: float64


## Analysis Conclusion: 
- Similarity - When comparing <font color=lightblue>downtown Toronto</font> with <font color=pink>downtown Vancouver</font>, the No.1 most common venue is <font color=brown>Coffee Shop</font>  for both districts (65% vs. 18%). 

- Dissimilarity - In <font color=pink>Downtown Vancouver</font>, Hotel (12%) and Pizza Place (12%) are among the most common venues other than <font color=brown>Coffee Shop</font>. In <font color=lightblue>downtown Toronto</font>, Cafe (12%) is among the most common venues other than <font color=brown>Coffee Shop</font>.